File: Java OOP/Homework06/presenters/CompanyPresenter.java
package presenters;

import models.Company;
import models.Department;
import models.Designer;
import models.Programmer;
import views.View;

public class CompanyPresenter {
    Company company;
    View view;

    public CompanyPresenter(Company company, View view) {
        this.company = company;
        this.view = view;
    }

    public int getMenuItem() {
        view.printMessage("1). Все отделы");
        view.printMessage("2). Выбрать отдел");
        view.printMessage("3). В отдел");
        view.printMessage("4). Создать отдел");
        view.printMessage("5). Удалить отдел");
        view.printMessage("0 - Выход");

        return Integer.parseInt(view.getInput());
    }

    public void menuControl() {
        int index;
        int menuIndex;

        do {            
            menuIndex = getMenuItem();
            switch (menuIndex) {
                case 1: // Все отделы
                    printDepartamentsInfo();
                    break;

                case 2: // Информация об отделе
                    printDepartamentInfo();
                    break;

                case 3: // В отдел
                    index = getIndex();
                    DeportamentPresenter deportamentPresenter = new DeportamentPresenter(company.getDepartment(index), view);
                    deportamentPresenter.menuControl();
                    break;

                case 4: // Создать отдел
                    addDepartment();
                    break;

                case 5: // Удалить отдел
                    removeDepartment();
                    break;

                case 0: // Удалить отдел
                    view.printMessage("Удачного дня!");
                    break;

                default:
                    view.printMessage("Некорректный ввод");
                    break;
            }
        } while (menuIndex != 0);
    }

    public void printDepartamentInfo() {
        view.printMessage("Введите номер департамента");
        int index = Integer.parseInt(view.getInput());
        view.printMessage(String.format("Отдел \"%s\"", company.getDepartments().get(index).getName()));
    }

    public void printDepartamentsInfo() {
        int index = 0;
        for (Department department : company.getDepartments()) {
            view.printMessage(String.format("[%d] Отдел \"%s\"", index, department.getName()));
            index++;
        }
    } 
    
    public void addDepartment() {   
        view.printMessage("Введите название отдела:");
        String name = view.getInput();

        view.printMessage("Вид отдела:\n1). Дизайн.\n2). Программирование\n>>> ");
        int res = Integer.parseInt(view.getInput());

        switch (res) {
            case 1:
                company.add(new Department<Designer>(name, "Designer"));

            case 2:
                company.add(new Department<Programmer>(name, "Programmer"));

            default:
            view.printMessage("Неверный ввод");
        }
    }

    public void removeDepartment() {
        int index = getIndex();
        company.remove(index);
    }

    public int getIndex() {
        view.printMessage("Введите номер отдела");
        return Integer.parseInt(view.getInput());
    }

    
}

File: Java OOP/Homework07/presenters/CompanyPresenter.py
from models.Company import Company
from models.Department import Department
from presenters.DepartmentPresenter import DepartmentPresenter
from views.View import View


class CompanyPresenter:
    def __init__(self, company: Company, view: View):
        self.company = company
        self.view = view

    def get_menu_item(self):
        self.view.print_message("1). Все отделы")
        self.view.print_message("2). Выбрать отдел")
        self.view.print_message("3). В отдел")
        self.view.print_message("4). Создать отдел")
        self.view.print_message("5). Удалить отдел")
        self.view.print_message("0 - Выход")

        return int(self.view.get_input())

    def menu_control(self):
        index = 0
        menu_index = 0

        menu_index = self.get_menu_item()
        while menu_index != 0:
            if menu_index == 1:
                self.print_departments_info()
            elif menu_index == 2:
                self.print_department_info()
            elif menu_index == 3:
                index = self.get_index()
                department_presenter = DepartmentPresenter(self.company.get_department(index), self.view)
                department_presenter.menu_control()
            elif menu_index == 4:
                self.add_department()
            elif menu_index == 5:
                self.remove_department()
            elif menu_index == 0:
                self.view.print_message("Удачного дня!")
            else:
                self.view.print_message("Некорректный ввод")

            menu_index = self.get_menu_item()

    def print_department_info(self):
        self.view.print_message("Введите номер департамента")
        index = int(self.view.get_input())
        self.view.print_message(f'Отдел "{self.company.departments[index].name}"')

    def print_departments_info(self):
        index = 0
        for department in self.company.departments:
            self.view.print_message(f'[{index}] Отдел "{department.name}"')
            index += 1

    def add_department(self):
        self.view.print_message("Введите название отдела:")
        name = self.view.get_input()

        self.view.print_message("Вид отдела:\n1). Дизайн.\n2). Программирование\n>>> ")
        res = int(self.view.get_input())

        if res == 1:
            self.company.add(Department(name, "Designer"))
        elif res == 2:
            self.company.add(Department(name, "Programmer"))
        else:
            self.view.print_message("Неверный ввод")

    def remove_department(self):
        index = self.get_index()
        self.company.remove(index)

    def get_index(self):
        self.view.print_message("Введите номер отдела")
        return int(self.view.get_input())
